### PR TITLE
[5.1] Fix Form::renderField() can cause duplicate database queries

### PR DESCRIPTION
**Pull request by Fedik**
Pull Request for Issue 
- #28777
- #13483

### Summary of Changes

Method `FormField::getLayoutData()` called multiple times while field rendering, which can cause a multiple SQL queries.
The PR introduces a new method for data collection with in memory caching `collectLayoutData()`. 

**Thing to consider:**
~~I marked `getLayoutData()` as deprecated, hovewer both methods can live side by side, if people like.~~
Undeprecated back :)

### Testing Instructions
Create a custom field under `libraries/src/Form/Field/TestField.php`
With following code:
```php
<?php
namespace Joomla\CMS\Form\Field;

use Joomla\CMS\Factory;

class TestField extends CheckboxesField
{
    protected $type = 'Test';

    protected function getOptions()
    {
        $db = Factory::getDbo();
        $db->setQuery(
            $db->getQuery(true)
                ->select($db->quoteName(['title', 'id'], ['text', 'value']))
                ->from('#__content')
                ->setLimit(5)
        );

        $list = array_map(function($r) {$r->checked = null;return $r; }, $db->loadObjectList());

        return array_merge(parent::getOptions(), $list);
    }
}
```

Then add this field in to mod_custom parameters XML: `modules/mod_custom/mod_custom.xml`
```xml
<field type="test" name="test" label="test"/>
<field type="test" name="test2" label="test2"/>
<field type="test" name="test3" label="test3"/>
```
Enable Debug, and Debug Query.
Go to edit the "Modules: Custom" parameters, and watch amount of queries in debug bar.

### Actual result BEFORE applying this Pull Request
let say will be 52 queries

### Expected result AFTER applying this Pull Request
Will be 6 queries less, so 46.

In my test I got 3 queries per 1 testing field.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/217
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by HLeithner** [comment]
If I read it correctly you introduced caching for all formfields right?

On comment on `boolean` parameters, based on who believe it's not the a good practice to use them (ex. https://understandlegacycode.com/blog/what-is-wrong-with-boolean-parameters/). I would suggest to keep getLayoutData and add the caching functionality, beside that add a `resetLayoutData()` (or similar name) function instead of adding a `boolean` parameter.

I'm not sure about the side effects on all fields, but it's possible that it breaks formfields which expect to be called several times? I'm not sure didn't thought about this long enough.

File: libraries/src/Form/Field/HiddenField.php [hunk 1]
```diff
@@ -39,29 +39,4 @@ class HiddenField extends FormField
...
-     * Method to get the field input markup.
-     *
-     * @return  string  The field input markup.
-     *
-     * @since   1.7.0
-     */
-    protected function getInput()
-    {
-        // Trim the trailing line in the layout file
-        return rtrim($this->getRenderer($this->layout)->render($this->getLayoutData()), PHP_EOL);
```
Comment (HLeithner) [inline libraries/src/Form/Field/HiddenField.php:53]: what's the reason for deleting it here but not in the other `getInput()` calls for example `EmailField.php`?

Comment (Fedik) [reply libraries/src/Form/Field/HiddenField.php:53]: There many other fields where this overrdie does not make any sense.
I did not wanted to putt all in here, maybe will do another PR some time later.

**Comment by Fedik** [comment]
> On comment on boolean parameters, based on who believe it's not the a good practice to use them

It just 1, not many of them (as in linked article) :) 
Hovewer, I can do without this parameter, and developers can reset the data property manualy.


> I'm not sure about the side effects on all fields, but it's possible that it breaks formfields which expect to be called several times?

I can't think any reason why it should be valid. The getData() should return consistent data, in any amount of call.

**Comment by Fedik** [comment]
One side effect, that many fields will work much faster, in a couple ns :wink:

**Comment by Fedik** [comment]
I have removed `reset` parameter

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 8d3650c0bc4aa2a33ee7bece10a898029f5f7b69<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42709">issues.joomla.org/tracker/joomla-cms/42709</a>.</sub>

**Comment by martin-zw** [comment]
I have tested this item :white_check_mark: successfully on 8d3650c0bc4aa2a33ee7bece10a898029f5f7b69<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42709">issues.joomla.org/tracker/joomla-cms/42709</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42709">issues.joomla.org/tracker/joomla-cms/42709</a>.</sub>

**Comment by LadySolveig** [comment]
Thank you @Fedik and for testing @alikon @martin-zw

**Comment by Fedik** [comment]
That sounds like a bug in the calendar field, I will make a fix later.

**Comment by Fedik** [comment]
@svenbluege please tets the fix https://github.com/joomla/joomla-cms/pull/43234

**Comment by dawe78** [comment]
Hi,
I got an issue with subforms, caused by this feature. In subforms, fields will be called multiple times; in this case, joomla 5.1 uses extra data (collected from getLayoutData()) from first field for all fields of same type, such as field value. This is a big issue and crashes all subforms!

It is a good thing to reduce database queries and use collected data for multiple calls just one time, but in this case, is causes a lot of issues.

Regards, Daniel<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42709">issues.joomla.org/tracker/joomla-cms/42709</a>.</sub>

**Comment by Fedik** [comment]
Please provide an example, how to reproduce that. Thanks.
And open a new issue for that.

**Pull request merged**